### ISSUE TITLE
Better stack traces

### ISSUE THREAD

**Issue by tolmasky** [ramda/ramda#2073]
So we've just shipped new stack traces in RunKit, and I'd really like Ramda to have easier to read stack traces (even outside of our tools). Currently it contains lots of Object.f2s and such, and I did a few experiments with setting functions names.

Here you can see what Ramda looks like currently:

<img width="940" alt="screen shot 2017-02-01 at 9 31 39 pm" src="https://cloud.githubusercontent.com/assets/23753/22538310/47ccb1fc-e8c6-11e6-902f-3234f076d6fd.png">

( https://runkit.com/tolmasky/ramda-current-traces )

And here is what they could look like:

<img width="976" alt="screen shot 2017-02-01 at 9 31 47 pm" src="https://cloud.githubusercontent.com/assets/23753/22538317/4f05ed30-e8c6-11e6-8ce6-4bf9662fb23b.png">

( https://runkit.com/tolmasky/ramda-better-traces-22 )

Basically, the trick is to do something like this:

```javascript
function functionNamed(aFunction, aName)
{
    Object.defineProperty(aFunction, "name", { value: aName });
    aFunction.displayName = aName;

    return aFunction;
}
```

**Comment by CrossEye** [comment]
I was hot on this some time back, using techniques from https://medium.com/@drboolean/debugging-functional-7deb4688a08c#.f0g8hif9c but I never got finished with it.  I would love to see something like this in the library, as much for debugging consoles as for stack traces.

It's a fair bit of work, but I think it would be worth it.

**Comment by agentultra** [comment]
I'd like to take a stab at this one.

**Comment by CrossEye** [comment]
Feel free.  I absolutely think it is worth doing.

**Comment by adrian-gierakowski** [comment]
@CrossEye I think stuff you mentions [here](https://github.com/ramda/ramda/pull/2082#issuecomment-278745116) regarding your ideas on improving `curry`, `partial`, `flip`, `arity`, `binary` makes total sense. Lodash is actually doing something of that sort already ( [see](https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L5527) ). They are using it for `bind`, `bindKey`, `partial`, `partialRigth`, `curry`, `curryRight`, `flip` and `rearg`. When you console.log a function wrapped in _.curry you might see something like:
```js
function (a, b) {
/* [wrapped with _.curry] */
    return a;
  }
```

I would also be inclined to work on this for ramda. Would you consider an implementation which ditches the placeholder in favour or rearg? Would you also might sharing some results of your experiments? What other ramda functions apart from  `curry`, `partial`, `flip`, `arity`, `binary` ( and their internal counterparts ) should be included in the refactor? thanks!

**Comment by buzzdecafe** [comment]
> Would you consider an implementation which ditches the placeholder in favour or rearg? 

Ditch the placeholder: :+1:
rearg: :-1:
